### PR TITLE
Avoid manual step for creating array setters

### PR DESCRIPTION
**Pull request by phanimarupaka**
@mortent 

This PR is to avoid additional manual step to create-setter for array types. 

Related issue: https://github.com/GoogleContainerTools/kpt/issues/698

[Here](https://googlecontainertools.github.io/kpt/guides/producer/setters/#setting-lists) is the current way to create setter and set values to array types. After creating the setter, user should manually move the setter ref to top level node for each setter ref which is a poor user experience. 

By this change, the `set` operation remains as is, but create-setter workflow is simplified. Here is the usage of command

```
# example.yaml
apiVersion: example.com/v1beta1
kind: Example
spec:
  list:
  - a
  - b
```

`kpt cfg create-setter --type array . list --field spec.list`

```
# example.yaml
apiVersion: example.com/v1beta1
kind: Example
spec:
  list: # {"$kpt-set":"list"}
  - a
  - b
```

`kpt cfg list-setters . list`

```
  NAME   VALUE   SET BY   DESCRIPTION   COUNT  
  list   [a,b]                          1      
```

`kpt cfg set . list c d`

```
# example.yaml
apiVersion: example.com/v1beta1
kind: Example
spec:
  list: # {"$kpt-set":"list"}
  - c
  - d
```

`kpt cfg list-setters . list`

```
  NAME   VALUE   SET BY   DESCRIPTION   COUNT  
  list   [c,d]                          1      
```

### CODE REVIEW

File: cmd/config/internal/commands/cmdlistsetters.go [hunk 1]
```diff
@@ -113,17 +113,20 @@ func (r *ListSettersRunner) ListSubstitutions(c *cobra.Command, args []string) e
 		return err
 	}
 	table := newTable(c.OutOrStdout(), r.Markdown)
-	table.SetHeader([]string{"SUBSTITUTION", "PATTERN", "SETTERS"})
+	table.SetHeader([]string{"SUBSTITUTION", "PATTERN", "REFERENCES"})
```
Comment (phanimarupaka) [inline cmd/config/internal/commands/cmdlistsetters.go:116]: As substitutions can refer to other substitutions, changing name to REFERENCES.

**Comment by mortent** [comment]
Can you add a description in the PR of how things are supposed to work after this change? I mostly understand the problem based on the reading the referenced issue, but it is now clear how we address that in this PR. This would give me better context when reading the code.

**Comment by mortent** [comment]
So the behavior of the setter after this change seems to be slightly different than our existing setters in that it doesn't require the user to specify the value. Instead we How would this change handle situations where the field might exist on more than one resource but with different list values?

File: cmd/config/internal/commands/cmdlistsetters_test.go [hunk 2]
```diff
@@ -251,8 +251,116 @@ spec:
...
+    io.k8s.cli.setters.list:
+      items:
+        type: string
+      maxItems: 3
+      type: array
+      description: hello world
+      x-k8s-cli:
+        setter:
+          name: list
+          value: List Values
```
Comment (mortent) [inline cmd/config/internal/commands/cmdlistsetters_test.go:275]: Why would we have both `value` and `listValues` set?

File: cmd/config/internal/commands/cmdcreatesetter.go [hunk 3]
```diff
@@ -124,6 +126,12 @@ func (r *CreateSetterRunner) preRunE(c *cobra.Command, args []string) error {
 		r.CreateSetter.Description = r.Set.SetPartialField.Description
 		r.CreateSetter.SetBy = r.Set.SetPartialField.SetBy
 		r.CreateSetter.Type = r.Set.SetPartialField.Type
+
+		if r.CreateSetter.Type == "array" {
+			// this is just a place holder value, we derive the actual value at
+			// later point from the field path
+			r.CreateSetter.FieldValue = "listValues"
```
[the remark this answers is not in the document]
Comment (phanimarupaka) [reply cmd/config/internal/commands/cmdcreatesetter.go]: Was using this string for setter to not match with any of the scalar values. It was a bit hacky. Set it to empty string instead and updated the logic.

File: cmd/config/internal/commands/cmdcreatesetter_test.go [hunk 4]
```diff
@@ -130,14 +130,16 @@ spec:
 
 		{
 			name:   "add replicas with schema list values",
-			args:   []string{"list", "a", "--description", "hello world", "--set-by", "me", "--type", "array"},
-			schema: `{"maxItems": 2, "type": "array", "items": {"type": "string"}}`,
+			args:   []string{"list", "a", "--description", "hello world", "--set-by", "me", "--type", "array", "--field", "spec.list"},
```
[the remark this answers is not in the document]
Comment (phanimarupaka) [reply cmd/config/internal/commands/cmdcreatesetter_test.go]: a is filed value. Which is not needed in array context. Updated the test. Even if user passes a value, we ignore it and derive the array values.

File: cmd/config/internal/commands/cmdlistsetters_test.go (hunk 2 above)
Comment (phanimarupaka) [reply cmd/config/internal/commands/cmdlistsetters_test.go]: Not needed. Removed.

File: cmd/config/internal/commands/cmdcreatesetter.go [hunk 5]
```diff
@@ -124,6 +126,16 @@ func (r *CreateSetterRunner) preRunE(c *cobra.Command, args []string) error {
...
 		r.CreateSetter.Type = r.Set.SetPartialField.Type
+
+		if r.CreateSetter.Type == "array" {
+			if !c.Flag("field").Changed {
+				return errors.Errorf("field flag must be set for array type setters")
+			}
+
+			// set it to empty string as the type is array and we derive the values at later point
+			// from filed path
+			r.CreateSetter.FieldValue = ""
```
Comment (mortent) [inline cmd/config/internal/commands/cmdcreatesetter.go:137]: This shouldn't be needed since empty string is the nil value for `FieldValue`.

Comment (phanimarupaka) [reply cmd/config/internal/commands/cmdcreatesetter.go:137]: The aim is to set it to empty even it user specifies some value as input.

File: kyaml/setters2/add.go [hunk 6]
```diff
@@ -48,17 +56,65 @@ func (a *Add) visitSequence(_ *yaml.RNode, _ string, _ *openapi.ResourceSchema)
...
+
+		// derive the list values for the sequence node to write it to openAPI definitions
+		var values []string
+		for _, sc := range node.Value.Content() {
+			values = append(values, sc.Value)
+		}
+
+		// check if there are different values for field path
+		if len(a.ListValues) > 0 {
+			sort.Strings(values)
```
Comment (mortent) [inline kyaml/setters2/add.go:81]: I don't think we should do this. Sequences are ordered in yaml, so it is not clear that the same values in a different order should be considered equal.

Comment (phanimarupaka) [reply kyaml/setters2/add.go]: Done.

**Comment by irvifa** [comment]
I wonder if this also can be used for array with key and value type as well

**Pull request merged**